File: services/report/internal/handler/csaf.go
package handler

import (
	"encoding/json"
	"net/http"

	"github.com/axebom/axebom/libs/go-shared/auth"
	"github.com/axebom/axebom/libs/go-shared/platform/errs"
	"github.com/axebom/axebom/services/report/internal/store"
)

// ---------------------------------------------------------------------------
// CSAF advisories — GET/POST /v1/csaf/{projectId}/advisories.
//
// ⚠ CSAF FOLLOWS VEX. A generate request names an EXISTING vex_statement_id;
// this handler never accepts a status/justification/scope directly — those
// came from scan-orchestrator's VEX write path already, and re-typing them
// here would let a client publish an advisory that disagrees with the VEX
// record behind it.
// ---------------------------------------------------------------------------

type csafAdvisoryDTO struct {
	ID             string          `json:"id"`
	VEXStatementID string          `json:"vex_statement_id"`
	TrackingID     string          `json:"tracking_id"`
	Description    string          `json:"description,omitempty"`
	Severity       string          `json:"severity,omitempty"`
	MitigationStep string          `json:"mitigation_steps,omitempty"`
	PublishedAt    string          `json:"published_at,omitempty"`
	Document       json.RawMessage `json:"document"`
	CreatedAt      string          `json:"created_at"`
}

func toCSAFAdvisoryDTO(a store.CSAFAdvisory) csafAdvisoryDTO {
	return csafAdvisoryDTO{
		ID: a.ID, VEXStatementID: a.VEXStatementID, TrackingID: a.TrackingID,
		Description: a.Description, Severity: a.Severity, MitigationStep: a.MitigationStep,
		PublishedAt: a.PublishedAt, Document: a.Document, CreatedAt: a.CreatedAt,
	}
}

type generateCSAFAdvisoryDTO struct {
	VEXStatementID   string   `json:"vex_statement_id"`
	ClusterDisplayID string   `json:"cluster_display_id"`
	ClusterAliases   []string `json:"cluster_aliases"`
	ComponentName    string   `json:"component_name"`
	ComponentPURL    string   `json:"component_purl"`
}

// GenerateCSAFAdvisory handles POST /v1/csaf/{projectId}/advisories.
func (h *Handler) GenerateCSAFAdvisory(w http.ResponseWriter, r *http.Request) {
	tenantID, err := auth.RequireTenant(r.Context())
	if err != nil {
		errs.Write(w, r, err)
		return
	}

	var dto generateCSAFAdvisoryDTO
	if err := decode(r, &dto); err != nil {
		errs.Write(w, r, err)
		return
	}
	if dto.VEXStatementID == "" {
		errs.Write(w, r, errs.New(errs.ValidationFieldRequired, "vex_statement_id is required"))
		return
	}

	advisory, err := h.svc.GenerateCSAFAdvisory(r.Context(), tenantID, r.PathValue("projectId"),
		store.GenerateCSAFAdvisoryInput{
			VEXStatementID:   dto.VEXStatementID,
			ClusterDisplayID: dto.ClusterDisplayID,
			ClusterAliases:   dto.ClusterAliases,
			ComponentName:    dto.ComponentName,
			ComponentPURL:    dto.ComponentPURL,
		}, h.now().UTC())
	if err != nil {
		errs.Write(w, r, err)
		return
	}
	errs.WriteJSON(w, http.StatusCreated, toCSAFAdvisoryDTO(*advisory))
}

// ListCSAFAdvisories handles GET /v1/csaf/{projectId}/advisories.
func (h *Handler) ListCSAFAdvisories(w http.ResponseWriter, r *http.Request) {
	tenantID, err := auth.RequireTenant(r.Context())
	if err != nil {
		errs.Write(w, r, err)
		return
	}

	advisories, err := h.svc.ListCSAFAdvisories(r.Context(), tenantID, r.PathValue("projectId"))
	if err != nil {
		errs.Write(w, r, err)
		return
	}

	out := make([]csafAdvisoryDTO, 0, len(advisories))
	for _, a := range advisories {
		out = append(out, toCSAFAdvisoryDTO(a))
	}
	errs.WriteJSON(w, http.StatusOK, map[string]any{"advisories": out})
}

File: frontend/src/lib/csaf.ts
/**
 * CSAF 2.0 advisories — POST/GET /v1/csaf/{projectId}/advisories.
 *
 * ⚠ CSAF FOLLOWS VEX. A generate request names an EXISTING vex_statement_id
 * — there is no way to publish an advisory that was not already triaged,
 * matching CERT-In's own discovery -> VEX -> CSAF sequence (p.35 Figure 7).
 * Generating twice for the same statement returns the same advisory rather
 * than publishing a duplicate.
 */

import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { api, request } from './api';

export interface CSAFAdvisory {
  id: string;
  vex_statement_id: string;
  tracking_id: string;
  description?: string;
  severity?: string;
  mitigation_steps?: string;
  published_at?: string;
  document: Record<string, unknown>;
  created_at: string;
}

export interface GenerateCSAFAdvisoryInput {
  vex_statement_id: string;
  cluster_display_id: string;
  cluster_aliases: string[];
  component_name: string;
  component_purl: string;
}

export function useCSAFAdvisories(projectId: string | undefined) {
  return useQuery({
    queryKey: ['csaf-advisories', projectId],
    queryFn: ({ signal }) =>
      request<{ advisories: CSAFAdvisory[] }>(`/v1/csaf/${projectId}/advisories`, { signal }),
    enabled: Boolean(projectId),
  });
}

export function useGenerateCSAFAdvisory(projectId: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (input: GenerateCSAFAdvisoryInput) =>
      api.post<CSAFAdvisory>(`/v1/csaf/${projectId}/advisories`, input),
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: ['csaf-advisories', projectId] });
    },
  });
}
